src/main/java/seedu/address/model/tut/Tutorial.java

```diff
@@ -7,186 +7,363 @@
+            requireNonNull(date);
+            requireNonNull(target);
+            TutDate tutDate = tutDates.computeIfAbsent(date, TutDate::new);
+            return students.stream()
+                    .filter(s -> s.getStudentId().equals(target))
+                    .findFirst()
+                    .map(student -> {
+                        student.setAttendance(tutDate);
+                        tutDate.add(target);
+                        return true;
+                    }).orElse(false);
+        }
+
+        @Override
+        public TutName getTutName() {
+            return this.tutName;
+        }
+
+        @Override
+        public boolean tutorialDateInList(TutDate date) {
+            return tutDates.containsKey(date.getDate());
+        }
+
+        @Override
+        public void addTutorialDate(TutDate tutorialDate) {
+            requireNonNull(tutorialDate);
+            tutDates.putIfAbsent(tutorialDate.getDate(), tutorialDate);
+        }
+
+        @Override
+        public TutDate getTutorialDate(Date date) {
+            return tutDates.get(date);
+        }
+
+        @Override
+        public boolean isValidTutorialDate(TutDate tutorialDate) {
+            return tutorialDate.isValid();
+        }
+
+        @Override
+        public List<TutDate> getTutDates() {
+            return new ArrayList<>(tutDates.values());
+        }
+
+        @Override
+        public TutorialClass getTutorialClass() {
+            return tutorialClass;
+        }
+
+        @Override
+        public void markAttendance(Student student, TutDate tutorialDate) throws TutDateNotFoundException {
+            if (!isValidTutorialDate(tutorialDate)) {
+                throw new TutDateNotFoundException();
+            }
+            if (!studentInList(student)) {
+                add(student);
+            }
+            addTutorialDate(tutorialDate);
+            tutorialDate.add(student.getStudentId());
+        }
+
+        @Override
+        public boolean studentInList(Student student) {
+            return students.contains(student);
+        }
+
+        @Override
+        public void deleteStudent(Student student) {
+            students.remove(student);
+        }
+
+        @Override
+        public boolean equals(Object other) {
+            if (other == this) {
+                return true;
+            }
+
+            if (other instanceof Exist) {
+                Exist exist = (Exist) other;
+
+                if (this.tutorialClass == exist.tutorialClass) {
+                    return true;
+                }
+
+                if (this.tutorialClass == null || exist.tutorialClass == null) {
+                    return false;
+                }
+                return tutorialClass.equals(exist.tutorialClass);
+            }
+
+            return false;
+        }
+        @Override
+        public String toString() {
+            return tutName + ": Tutorial " + tutorialClass;
+        }
```

Comment:
Nice reimplementation for Tutorial class to ensure when deleteTut command get executed no null class exception will be thrown

Reply:
Yes I do agree! Good job in making use of polymorphism which fits the OOP paradigm.